File: Depth_first_search_algorithm/Bài_toán_13.cpp
#include <iostream>
#include <vector>
using namespace std;

void dfs_component(int u, const vector<vector<int>>& adj, vector<bool>& visited, vector<int>& comp) {
    visited[u] = true;
    comp.push_back(u);
    for (int v : adj[u]) {
        if (!visited[v])
            dfs_component(v, adj, visited, comp);
    }
}

void general_dfs(const vector<vector<int>>& adj) {
    int n = adj.size();
    vector<bool> visited(n, false);
    int comp_id = 1;

    for (int i = 0; i < n; ++i) {
        if (!visited[i]) {
            vector<int> comp;
            dfs_component(i, adj, visited, comp);
            cout << "Component " << comp_id++ << ": ";
            for (int v : comp) cout << v << " ";
            cout << "\n";
        }
    }
}

int main() {
    vector<vector<int>> graph = {
        {1, 1},
        {1, 2},
        {},
        {5},
        {4}
    };

    general_dfs(graph);
    return 0;
}

File: Depth_first_search_algorithm/Bài_toán_13.py
def dfs_component(u, adj, visited, component):
    visited.add(u)
    component.append(u)
    for v in adj[u]:
        if v not in visited:
            dfs_component(v, adj, visited, component)

def general_dfs(adj):
    n = len(adj)
    visited = set()
    components = []

    for u in range(n):
        if u not in visited:
            comp = []
            dfs_component(u, adj, visited, comp)
            components.append(comp)
    return components

graph = {
    0: [1, 1],
    1: [0, 2],
    2: [1, 2],
    3: [],
    4: [5],
    5: [4]
}

result = general_dfs(graph)
for i, comp in enumerate(result):
    print(f"Component {i+1}:", comp)
